lgc/tool/lgc/lgc.cpp

```diff
@@ -0,0 +1,259 @@
+          // End the current split module and go on to the next one.
+          separatedAsms.back() = separatedAsms.back().slice(0, remaining.data() - separatedAsms.back().data());
+          separatedAsms.push_back(remaining);
+          hadMetadata = false;
+        }
+      }
+      auto nlPos = remaining.find_first_of('\n');
+      if (nlPos == StringRef::npos)
+        break;
+      remaining = remaining.slice(nlPos + 1, StringRef::npos);
+    }
+
+    // Check that the -extract option is not out of range.
+    if (Extract > separatedAsms.size()) {
+      errs() << progName << ": " << bufferName << ": Not enough modules for -extract value\n";
+      exit(1);
+    }
+
+    // Process each module. Put extra newlines at the start of each one other than the first so that
+    // line numbers are correct for error reporting.
+    unsigned extraNlCount = 0;
+    for (unsigned idx = 0; idx != separatedAsms.size(); ++idx) {
+      StringRef separatedAsm = separatedAsms[idx];
+      std::string asmText;
+      asmText.insert(asmText.end(), extraNlCount, '\n');
+      extraNlCount += separatedAsm.count('\n');
+      asmText += separatedAsm;
+
+      // Skip this module if -extract was specified for a different index.
+      if (Extract && Extract != idx + 1)
+        continue;
+
+      // Use a MemoryBufferRef with the original filename so error reporting reports it.
+      MemoryBufferRef asmBuffer(asmText, bufferName);
+
+      // Assemble the text
+      SMDiagnostic error;
+      std::unique_ptr<Module> module = parseAssembly(asmBuffer, error, context);
+      if (!module) {
+        error.print(progName, errs());
+        errs() << "\n";
+        return 1;
+      }
+
+      // Verify the resulting IR.
+      if (verifyModule(*module, &errs())) {
+        errs() << progName << ": " << bufferName << ": IR verification errors in module " << idx << "\n";
+        return 1;
+      }
+
+      // Determine whether we are outputting to a file.
+      bool outputToFile = false;
+      if (OutFileName.empty()) {
+        // No -o specified: output to stdout if input is -
+        outputToFile = inFileName != "-";
+      }
+
+      // Create a Pipeline and run the middle-end compiler.
+      SmallString<16> outBuffer;
+      raw_svector_ostream outStream(outBuffer);
+      std::unique_ptr<Pipeline> pipeline(lgcContext->createPipeline());
+      pipeline->generate(std::move(module), outStream, nullptr, {});
+
+      // Output to stdout if applicable.
+      if (outputToFile == false) {
+        outs() << outBuffer;
+        continue;
+      }
+
+      SmallString<64> outFileName(OutFileName);
+      if (outFileName.empty()) {
+        // Determine the output filename by taking the input filename, removing the directory,
+        // removing the extension, and adding a default extension that depends on the output contents.
+        const char *ext = ".s";
+        if (isElfBinary(outBuffer)) {
+          ext = ".elf";
```

Comment:
Can we have these extension names as top-level `llvm::StringLiteral` constants?

Reply:
I find the practice of putting string constants out-of-line makes code a lot harder to read, especially when they're perfectly well-known strings like ".s".

Though one piece of bike-shedding I'd have here is that if we follow the practice of calling assembly files ".s", why aren't we calling ELF files ".o"? ;)

Reply:
That would be nice, but I think we have to follow the driver's practice of calling a pal pipeline elf ".elf".

Maybe it would be nice to use ".o" for an unlinked compiled shader, before linking, but the problem with that is that the only way of telling the difference is to parse the elf and find some bit of pal metadata (that we haven't defined yet). Both kinds of elf are "relocatable" in elf standard terms.

Reply:
Yeah, it's not a big deal.

Reply:
Because PAL has an internal ELF analyzer that accepts ".elf" as the inputs. It is better to keep the output file extension consistent with this.

Reply:
Ack, fine for me.